### PR TITLE
Story/cite 186

### PR DESCRIPTION
**Pull request by agoyal61**
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
 

https://diging.atlassian.net/browse/CITE-186

... Put ticket description here and add link to ticket ...

Manage authority entries page needs to be paginated
Right now it shows all entries but it should be paginated so that it only shows 20 at a time.

## Are there any other pull requests that this one depends on?

  
  https://github.com/diging/citesphere-model/pull/41
  
## Anything else the reviewer needs to know?

... describe here ...

### CODE REVIEW

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AuthorityListController.java [hunk 1]
```diff
@@ -3,57 +3,105 @@
...
     
     @RequestMapping("/auth/authority/user/list")
-    public String getAuthoritiesForUser(Model model, Authentication authentication) {
+    public String getAuthoritiesForUser(Model model, Authentication authentication,
+            @RequestParam(defaultValue = "1", required = false, value = "page") String page) {
+        Integer pageInt = 1;
+        try {
+            pageInt = new Integer(page);
+        } catch (NumberFormatException ex) {
+            logger.warn("Trying to access invalid page number: " + page);
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AuthorityListController.java:93]: same as above

File: citesphere/src/main/webapp/WEB-INF/views/auth/authorities/list.html [hunk 2]
```diff
@@ -1,10 +1,30 @@
...
+  	$('#pagination-top').twbsPagination({
+        totalPages: [(${totalPages})],
+	    startPage: [(${currentPage})],
+	    prev: "«",
+	    next: "»",
+	    visiblePages: 10,
+	    initiateStartPageClick: false,
+        onPageClick: function (event, page) {
+     	    if($("#displayAuthorities").val() === 'all'){
+     	    	 console.log("all")
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/authorities/list.html:18]: remove

File: citesphere/src/main/webapp/WEB-INF/views/auth/authorities/list.html [hunk 3]
```diff
@@ -1,10 +1,30 @@
...
+     	    	 window.location.href = [[@{/auth/authority/list}]] + "?" + "page=" + page;
+    	    } else if($("#displayAuthorities").val() === 'userSpecific'){
+    	    	window.location.href = [[@{/auth/authority/user/list}]] + "?" + "page=" + page;
+    	    } else {
+    	    	var zoteroGroupId = $("#displayAuthorities").val();
+    	    	window.location.href = [[@{/auth/authority/}]] + zoteroGroupId + "/list" + "?" + "page=" + page;
+    	    }
```
Comment (jdamerow) [inline citesphere/src/main/webapp/WEB-INF/views/auth/authorities/list.html:25]: this looks like the same code as below in the `$("#displayAuthorities").change` function. Should be DRY.

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/AuthorityServiceTest.java [hunk 4]
```diff
@@ -593,4 +598,72 @@ public void test_searchAuthorityEntries_NoAuthorityFound()
...
+    
+    @Test()
+    public void test_getAuthoritiesByGroup_emptyResult() {
+        Mockito.when(entryRepository.findByGroupsOrderByName(groupId, paging)).thenReturn(new PageImpl<>(new ArrayList<>()));
+        Page<IAuthorityEntry> searchResult = managerToTest.getAuthoritiesByGroup(groupId, page, pageSize);
+        Assert.assertTrue(searchResult.isEmpty());
+    }
+    
+    @Test()
+    public void test_getUserSpecificAuthorities() {
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/AuthorityServiceTest.java:648]: same as above

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/AuthorityServiceTest.java [hunk 5]
```diff
@@ -593,4 +598,72 @@ public void test_searchAuthorityEntries_NoAuthorityFound()
...
+        entriesAlbert.add(entry2);
+        entriesAlbert.add(entry3);
+        Page<IAuthorityEntry> entries = new PageImpl<>(entriesAlbert);
+        Mockito.when(entryRepository.findByUsernameOrGroupsInOrderByName(username, groupIds, paging)).thenReturn(entries);
+        Page<IAuthorityEntry> searchResult = managerToTest.getAll(user, groupIds, page, pageSize);
+        Assert.assertEquals(2, searchResult.getContent().size());
+        for (IAuthorityEntry entry : searchResult) {
+            Assert.assertTrue(entry.getGroups().contains(groupId));
+            Assert.assertTrue(entry.getName().contains("Albert"));
+            Assert.assertTrue(!entry.getName().contains("Einstein"));
```
Comment (jdamerow) [inline citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/AuthorityServiceTest.java:613]: what are these asserts testing? the point is that all entries returned by the entryRepository are also returned by the manager, not that the names include "Albert" but not "Einstein".

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Make it so, Jenkins.

File: citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AuthorityListController.java [hunk 6]
```diff
@@ -3,57 +3,105 @@
...
-    public String getAuthoritiesForUser(Model model, Authentication authentication) {
+    public String getAuthoritiesForUser(Model model, Authentication authentication,
+            @RequestParam(defaultValue = "1", required = false, value = "page") String page) {
+        Integer pageInt = 1;
+        try {
+            pageInt = new Integer(page);
+        } catch (NumberFormatException ex) {
+            logger.error("Trying to access invalid page number: ", ex);
+        }
+        pageInt = (pageInt - 1) < 0 ? 0 : pageInt - 1;
```
Comment (jdamerow) [inline citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AuthorityListController.java:95]: see above

**Comment by jdamerow** [comment]
When I go to a group that doesn't have any entries, i get this error (and the drop down is broken afterwards):
```
jquery.twbsPagination.min.js:10 Uncaught Error: Start page option is incorrect
    at new r (jquery.twbsPagination.min.js:10:233)
    at o.fn.twbsPagination (jquery.twbsPagination.min.js:10:5530)
    at HTMLDocument.<anonymous> (list:57:25)
    at j (jquery.js:3148:30)
    at Object.fireWith [as resolveWith] (jquery.js:3260:7)
    at Function.ready (jquery.js:3472:13)
    at HTMLDocument.J (jquery.js:3503:10)
```

**Comment by agoyal61** [comment]
Added the condition for total page to handle pagination error

**Comment by jdamerow** [comment]
Make it so, Jenkins.

**Pull request merged**